src/main/java/com/google/devtools/build/lib/bazel/bzlmod/BazelLockFileModule.java

```diff
@@ -163,15 +246,18 @@ public void afterCommand() {
         updatedExtensionMap.put(moduleExtensionId, entry.getValue());
         continue;
       }
-      if (!newExtensionInfo.moduleExtension().shouldLockExtension()) {
-        // Extension has become reproducible and should not be locked anymore.
-        continue;
-      }
       var newFactors = newExtensionInfo.extensionFactors();
+      // Factor results can be individually marked as reproducible.
       ImmutableSortedMap<ModuleExtensionEvalFactors, LockFileModuleExtension>
           perFactorResultsToKeep =
               ImmutableSortedMap.copyOf(
-                  Maps.filterKeys(entry.getValue(), newFactors::hasSameDependenciesAs));
+                  Maps.filterKeys(
+                      entry.getValue(),
+                      existingFactors ->
```

Comment:
can this part be extracted into a static method or something? it's extremely dense and rather hard to understand.